external-crates/move/crates/move-analyzer/src/completion.rs

```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
+    cursor: &CursorContext,
+    mod_ident: &ModuleIdent,
+    chain_target: ChainCompletionTarget,
+) -> Vec<CompletionItem> {
+    use ChainCompletionTarget as CT;
+
+    let mut completions = vec![];
+
+    let Some(mod_defs) = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .find(|mdef| mdef.ident == mod_ident.value)
+    else {
+        return completions;
+    };
+
+    // list all members or only publicly visible ones
+    let mut same_module = false;
+    let mut same_package = false;
+    if let Some(cursor_mod_ident) = cursor.module {
+        if &cursor_mod_ident == mod_ident {
+            same_module = true;
+        }
+        if cursor_mod_ident.value.address == mod_ident.value.address {
+            same_package = true;
+        }
+    }
+
+    if matches!(chain_target, CT::Function) || matches!(chain_target, CT::All) {
+        for (fname, fdef) in &mod_defs.functions {
+            let Some(DefInfo::Function(
+                _,
+                visibility,
+                fun_type,
+                _,
+                type_args,
+                arg_names,
+                arg_types,
+                ret_type,
+                _,
+            )) = symbols.def_info(&fdef.name_loc)
+            else {
+                continue;
+            };
+
+            if matches!(visibility, Visibility::Internal) && !same_module {
+                continue;
+            }
+            if matches!(visibility, Visibility::Package(_)) && !same_package {
+                continue;
+            }
+
+            completions.push(call_completion_item(
+                &mod_ident.value,
+                matches!(fun_type, FunType::Macro),
+                None,
+                fname,
+                type_args,
+                arg_names,
+                arg_types,
+                ret_type,
+            ));
+        }
+    }
+
+    if matches!(chain_target, CT::Type) || matches!(chain_target, CT::All) {
+        for sname in mod_defs.structs.keys() {
+            completions.push(completion_item(sname, CompletionItemKind::STRUCT));
+        }
+    }
+
+    if matches!(chain_target, CT::Type) || matches!(chain_target, CT::All) {
+        for ename in mod_defs.enums.keys() {
+            completions.push(completion_item(ename, CompletionItemKind::ENUM));
+        }
+    }
+
+    if matches!(chain_target, CT::All) {
+        for cname in mod_defs.constants.keys() {
+            if !same_module {
+                continue;
+            }
+            completions.push(completion_item(cname, CompletionItemKind::CONSTANT));
+        }
+    }
+
+    completions
+}
+
+/// Returns completion item if a given name/alias identifies a valid member of a given module
+/// available in the completion scope
+fn first_position_member_completion(
+    symbols: &Symbols,
+    mod_ident: &ModuleIdent_,
+    member_alias: &Symbol,
+    member_name: &Symbol,
+    chain_target: ChainCompletionTarget,
+) -> Option<CompletionItem> {
```

Comment:
Because one is dealing with all available completions and the other checks if a given completion is valid, I don't really see an easy and clean way to share the code without resorting to constructs that I am not sure would actually be an improvement here (e.g., callbacks or macros)